File: BE/tests/HuyHieuDang.Web.IntegrationTests/ImportFixtures.cs
using System.Text.Json.Serialization;

namespace HuyHieuDang.Web.IntegrationTests;

/// <summary>
/// Trỏ tới bộ tệp Excel dùng chung của QC ở <c>tests/fixtures/excel/</c>. Bộ kiểm thử tích hợp
/// chạy qua chính những tệp này, không tự dựng tệp khác thay thế.
/// </summary>
public static class ImportFixtures
{
    /// <summary>
    /// Đường dẫn thư mục <c>tests/fixtures/excel</c> của kho mã.
    /// </summary>
    public static string ExcelDirectory { get; } = Locate();

    /// <summary>
    /// Lấy đường dẫn đầy đủ của một tệp fixture.
    /// </summary>
    /// <param name="fileName">Tên tệp, ví dụ <c>loi-4-dong.xlsx</c>.</param>
    /// <returns>Đường dẫn đầy đủ.</returns>
    public static string Path(string fileName) => System.IO.Path.Combine(ExcelDirectory, fileName);

    /// <summary>
    /// Đọc toàn bộ nội dung một tệp fixture.
    /// </summary>
    /// <param name="fileName">Tên tệp.</param>
    /// <returns>Nội dung nhị phân.</returns>
    public static byte[] Read(string fileName) => File.ReadAllBytes(Path(fileName));

    /// <summary>
    /// Đi ngược cây thư mục từ nơi chạy bản dựng cho tới khi gặp <c>tests/fixtures/excel</c>.
    /// </summary>
    /// <returns>Đường dẫn thư mục fixture.</returns>
    private static string Locate()
    {
        DirectoryInfo? directory = new(AppContext.BaseDirectory);

        while (directory is not null)
        {
            string candidate = System.IO.Path.Combine(directory.FullName, "tests", "fixtures", "excel");

            if (Directory.Exists(candidate))
            {
                return candidate;
            }

            directory = directory.Parent;
        }

        throw new DirectoryNotFoundException("Không tìm thấy thư mục tests/fixtures/excel trong kho mã.");
    }
}

/// <summary>
/// Phần <c>data</c> của <c>POST /api/PartyMembers/Import/Preview</c> (mục 4.2 hợp đồng API).
/// </summary>
public sealed class ImportPreviewPayload
{
    [JsonPropertyName("fileName")]
    public string? FileName { get; set; }

    [JsonPropertyName("totalRows")]
    public int TotalRows { get; set; }

    [JsonPropertyName("validCount")]
    public int ValidCount { get; set; }

    [JsonPropertyName("errorCount")]
    public int ErrorCount { get; set; }

    [JsonPropertyName("validRows")]
    public List<ImportValidRowPayload> ValidRows { get; set; } = new();

    [JsonPropertyName("errorRows")]
    public List<ImportErrorRowPayload> ErrorRows { get; set; } = new();
}

/// <summary>
/// Một dòng hợp lệ trong kết quả xem trước.
/// </summary>
public sealed class ImportValidRowPayload
{
    [JsonPropertyName("rowNumber")]
    public int RowNumber { get; set; }

    [JsonPropertyName("fullName")]
    public string? FullName { get; set; }

    [JsonPropertyName("dateOfBirth")]
    public DateOnly? DateOfBirth { get; set; }

    [JsonPropertyName("gender")]
    public string? Gender { get; set; }

    [JsonPropertyName("officialAdmissionDate")]
    public DateOnly OfficialAdmissionDate { get; set; }
}

/// <summary>
/// Một dòng lỗi trong kết quả xem trước; bốn trường giữ nguyên chữ thô.
/// </summary>
public sealed class ImportErrorRowPayload
{
    [JsonPropertyName("rowNumber")]
    public int RowNumber { get; set; }

    [JsonPropertyName("fullName")]
    public string? FullName { get; set; }

    [JsonPropertyName("dateOfBirth")]
    public string? DateOfBirth { get; set; }

    [JsonPropertyName("gender")]
    public string? Gender { get; set; }

    [JsonPropertyName("officialAdmissionDate")]
    public string? OfficialAdmissionDate { get; set; }

    [JsonPropertyName("errors")]
    public List<ImportRowErrorPayload> Errors { get; set; } = new();
}

/// <summary>
/// Một lý do của dòng lỗi.
/// </summary>
public sealed class ImportRowErrorPayload
{
    [JsonPropertyName("errorCode")]
    public string? ErrorCode { get; set; }

    [JsonPropertyName("field")]
    public string? Field { get; set; }
}

/// <summary>
/// Phần <c>data</c> của <c>POST /api/PartyMembers/Import/Commit</c> (mục 4.3 hợp đồng API).
/// </summary>
public sealed class ImportCommitPayload
{
    [JsonPropertyName("importedCount")]
    public int ImportedCount { get; set; }

    [JsonPropertyName("skippedCount")]
    public int SkippedCount { get; set; }
}

File: FE/src/mocks/importFixtures.ts
/**
 * Kết quả xem trước import giả — dựng từ `tests/fixtures/data/import-rows.json`
 * của QC, đã ánh xạ sang mã lỗi của hợp đồng API (mục 4.2, OQ-2).
 */
import type { ImportPreview } from '../api/imports';

export const MOCK_IMPORT_PREVIEW: ImportPreview = {
  fileName: 'DanhSachDangVien_MauKiemThu.xlsx',
  totalRows: 14,
  validCount: 6,
  errorCount: 8,
  validRows: [
    {
      rowNumber: 2,
      fullName: 'Nguyễn Hợp Lệ Một',
      dateOfBirth: '1974-03-12',
      gender: 'Male',
      officialAdmissionDate: '1996-10-01',
    },
    {
      rowNumber: 3,
      fullName: 'Trần Hợp Lệ Hai',
      dateOfBirth: '1973-07-05',
      gender: 'Female',
      officialAdmissionDate: '1996-11-07',
    },
    {
      rowNumber: 4,
      fullName: 'Lê Hợp Lệ Ba',
      dateOfBirth: null,
      gender: null,
      officialAdmissionDate: '1996-01-15',
    },
    {
      rowNumber: 5,
      fullName: 'Phạm Hợp Lệ Bốn',
      dateOfBirth: '1975-02-09',
      gender: 'Female',
      officialAdmissionDate: '1996-05-31',
    },
    {
      rowNumber: 6,
      fullName: 'Hoàng Hợp Lệ Năm',
      dateOfBirth: '1971-06-30',
      gender: 'Male',
      officialAdmissionDate: '1996-08-15',
    },
    {
      rowNumber: 7,
      fullName: 'Vũ Hợp Lệ Sáu',
      dateOfBirth: '1972-09-14',
      gender: 'Female',
      officialAdmissionDate: '1996-09-10',
    },
  ],
  errorRows: [
    {
      rowNumber: 8,
      fullName: '',
      dateOfBirth: '12/03/1974',
      gender: 'Nam',
      officialAdmissionDate: '01/10/1996',
      errors: [
        {
          errorCode: 'MissingFullName',
          field: 'FullName',
        },
      ],
    },
    {
      rowNumber: 9,
      fullName: 'Nguyễn Thiếu Ngày',
      dateOfBirth: '12/03/1974',
      gender: 'Nam',
      officialAdmissionDate: '',
      errors: [
        {
          errorCode: 'MissingOfficialAdmissionDate',
          field: 'OfficialAdmissionDate',
        },
      ],
    },
    {
      rowNumber: 10,
      fullName: 'Trần Sai Định Dạng',
      dateOfBirth: '12/03/1974',
      gender: 'Nữ',
      officialAdmissionDate: '1996-10-01',
      errors: [
        {
          errorCode: 'InvalidDateFormat',
          field: 'OfficialAdmissionDate',
        },
      ],
    },
    {
      rowNumber: 11,
      fullName: 'Lê Ngày Tương Lai',
      dateOfBirth: '12/03/1974',
      gender: 'Nam',
      officialAdmissionDate: '20/09/2026',
      errors: [
        {
          errorCode: 'FutureOfficialAdmissionDate',
          field: 'OfficialAdmissionDate',
        },
      ],
    },
    {
      rowNumber: 12,
      fullName: 'Phạm Giới Tính Lạ',
      dateOfBirth: '12/03/1974',
      gender: 'Khác',
      officialAdmissionDate: '01/10/1996',
      errors: [
        {
          errorCode: 'InvalidGender',
          field: 'Gender',
        },
      ],
    },
    {
      rowNumber: 13,
      fullName: 'Hoàng Sinh Sau',
      dateOfBirth: '02/01/1997',
      gender: 'Nam',
      officialAdmissionDate: '01/10/1996',
      errors: [
        {
          errorCode: 'BirthDateAfterAdmissionDate',
          field: 'DateOfBirth',
        },
      ],
    },
    {
      rowNumber: 14,
      fullName: 'Vũ Ngày Sinh Sai',
      dateOfBirth: '31/02/1974',
      gender: 'Nữ',
      officialAdmissionDate: '01/10/1996',
      errors: [
        {
          errorCode: 'InvalidDateFormat',
          field: 'DateOfBirth',
        },
      ],
    },
    {
      rowNumber: 15,
      fullName: '',
      dateOfBirth: '31/02/1974',
      gender: 'Khác',
      officialAdmissionDate: '',
      errors: [
        {
          errorCode: 'MissingFullName',
          field: 'FullName',
        },
        {
          errorCode: 'MissingOfficialAdmissionDate',
          field: 'OfficialAdmissionDate',
        },
        {
          errorCode: 'InvalidDateFormat',
          field: 'DateOfBirth',
        },
        {
          errorCode: 'InvalidGender',
          field: 'Gender',
        },
      ],
    },
  ],
};
